src/expr/impl/src/scalar/vector.rs

```diff
@@ -576,3 +576,107 @@ fn l2_normalize(vector: VectorRef<'_>) -> Result<VectorVal> {
         .map_err(|_| ExprError::NumericOverflow)?;
     Ok(result)
 }
+
+/// ```slt
+/// query R
+/// SELECT vector_dims('[1,2,3]'::vector(3));
+/// ----
+/// 3
+///
+/// query R
+/// SELECT vector_dims('[10,20,30,40]'::vector(4));
+/// ----
+/// 4
+///
+/// query error dimensions
+/// SELECT vector_dims('[]'::vector(0));
+/// ```
+#[function("vector_dims(vector) -> int4")]
+fn vector_dim(v: VectorRef<'_>) -> Result<i32> {
+    let vector = v.into_slice();
+    check_dim("vector_dim", vector.len() as i32)?;
+    Ok(vector.len() as i32)
+}
+
+fn check_dim(name: &'static str, dim: i32) -> Result<()> {
+    if dim < 1 {
+        return Err(ExprError::InvalidParam {
+            name,
+            reason: "vector must have at least 1 dimension".into(),
+        });
+    }
+    if dim > (DataType::VEC_MAX_SIZE as i32) {
+        return Err(ExprError::InvalidParam {
+            name,
+            reason: format!("vector cannot have more than {} dimensions", dim).into(),
+        });
+    }
+    Ok(())
+}
+
+/// ```slt
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector(5), 1, 3);
+/// ----
+/// [1,2,3]
+///
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector(5), 3, 2);
+/// ----
+/// [3, 4]
+///
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector(5), -1, 3);
+/// ----
+/// [1]
+///
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector(5), 3, 9);
+/// ----
+/// [3,4,5]
+///
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector, -2147483644, 2147483647);
+/// ----
+/// [1,2]
+///
+/// query R
+/// SELECT subvector('[1,2,3,4,5]'::vector, 3, 2147483647);
+/// ----
+/// [3,4,5]
+/// ```
+#[function("subvector(vector, int4, int4) -> vector", type_infer = "unreachable")]
+fn subvector(v: VectorRef<'_>, start: i32, count: i32) -> Result<VectorVal> {
```

Comment:
The dimension of a vector column should be statically inferable, because the data type of a column should be static to be `DataType::Vector(dimension)`. However, the parameter `start` and `count` is dynamic, and then the dimension is variable and cannot have a fixed dimension to form the fixed data type `DataType::Vector(dimension)`.

Reply:
Theoratical, if the value of `start` and `count` is statically constant, we can derive the dimension of the output vector.

I think in `bind_builtin_scalar_function`, when you handle `subvector `, you can check call `eval_const` on the two inputs to obtain constant `start` and `count`, and then derive the output dimension, and call `FunctionCall::new_unchecked` to specify the input and output data type directly.